From 25d6e2de3f0e4f3ebc54d8a7ebd20c9297f4a21a Mon Sep 17 00:00:00 2001
From: Ram Kumar Dwivedi <rdwivedi@qti.qualcomm.com>
Date: Tue, 1 Jul 2025 11:36:02 +0530
Subject: [PATCH 1/2] PENDING: ufs: ufs-qcom: Remove reset GPIO Dependency and
 Enable error handling

The reset GPIO is no longer required on firmware-managed platforms,
as the reset sequence is handled by the firmware itself. Removing
this dependency eliminates redundant hardware configuration.

Introduce error handling support for firmware-managed targets to
enable fault recovery in case of errors.

Refactor the code structure to enhance clarity and maintainability.

Change-Id: I0e6845543c637e17092e5a8cfd2d8cb643d08547
Signed-off-by: Ram Kumar Dwivedi <rdwivedi@qti.qualcomm.com>
---
 drivers/ufs/host/ufs-qcom.c | 150 +++++++++++++++++++++---------------
 1 file changed, 89 insertions(+), 61 deletions(-)

diff --git a/drivers/ufs/host/ufs-qcom.c b/drivers/ufs/host/ufs-qcom.c
index 65e6baa0a911..d63f337bc891 100644
--- a/drivers/ufs/host/ufs-qcom.c
+++ b/drivers/ufs/host/ufs-qcom.c
@@ -1434,14 +1434,24 @@ static int ufs_qcom_icc_init(struct ufs_qcom_host *host)
 	return 0;
 }
 
-static int ufs_qcom_common_init(struct ufs_hba *hba)
+/**
+ * ufs_qcom_init - bind phy with controller
+ * @hba: host controller instance
+ *
+ * Binds PHY with controller and powers up PHY enabling clocks
+ * and regulators.
+ *
+ * Return: -EPROBE_DEFER if binding fails, returns negative error
+ * on phy power up failure and returns zero on success.
+ */
+static int ufs_qcom_init(struct ufs_hba *hba)
 {
+	int err;
 	struct device *dev = hba->dev;
 	struct platform_device *pdev = to_platform_device(dev);
 	struct ufs_qcom_host *host;
-	struct ufs_clk_info *clki;
 	struct resource *res;
-	int err;
+	struct ufs_clk_info *clki;
 
 	host = devm_kzalloc(dev, sizeof(*host), GFP_KERNEL);
 	if (!host) {
@@ -1458,20 +1468,37 @@ static int ufs_qcom_common_init(struct ufs_hba *hba)
 	if (IS_ERR(host->core_reset)) {
 		err = dev_err_probe(dev, PTR_ERR(host->core_reset),
 				    "Failed to get reset control\n");
-		return err;
+		goto out_variant_clear;
+	}
+
+	/* Fire up the reset controller. Failure here is non-fatal. */
+	host->rcdev.of_node = dev->of_node;
+	host->rcdev.ops = &ufs_qcom_reset_ops;
+	host->rcdev.owner = dev->driver->owner;
+	host->rcdev.nr_resets = 1;
+	err = devm_reset_controller_register(dev, &host->rcdev);
+	if (err)
+		dev_warn(dev, "Failed to register reset controller\n");
+
+	if (!has_acpi_companion(dev)) {
+		host->generic_phy = devm_phy_get(dev, "ufsphy");
+		if (IS_ERR(host->generic_phy)) {
+			err = dev_err_probe(dev, PTR_ERR(host->generic_phy), "Failed to get PHY\n");
+			goto out_variant_clear;
+		}
 	}
 
 	err = ufs_qcom_icc_init(host);
 	if (err)
-		return err;
+		goto out_variant_clear;
 
 	host->device_reset = devm_gpiod_get_optional(dev, "reset",
-						     GPIOD_OUT_LOW);
+						     GPIOD_OUT_HIGH);
 	if (IS_ERR(host->device_reset)) {
 		err = PTR_ERR(host->device_reset);
 		if (err != -EPROBE_DEFER)
 			dev_err(dev, "failed to acquire reset gpio: %d\n", err);
-		return err;
+		goto out_variant_clear;
 	}
 
 	ufs_qcom_get_controller_revision(hba, &host->hw_ver.major,
@@ -1502,12 +1529,20 @@ static int ufs_qcom_common_init(struct ufs_hba *hba)
 			clki->keep_link_active = true;
 	}
 
+	err = ufs_qcom_init_lane_clks(host);
+	if (err)
+		goto out_variant_clear;
+
+	ufs_qcom_set_caps(hba);
+	ufs_qcom_advertise_quirks(hba);
 	ufs_qcom_set_host_params(hba);
 	ufs_qcom_set_phy_gear(host);
 
 	err = ufs_qcom_ice_init(host);
 	if (err)
-		return err;
+		goto out_variant_clear;
+
+	ufs_qcom_setup_clocks(hba, true, POST_CHANGE);
 
 	if (hba->dev->id < MAX_UFS_QCOM_HOSTS)
 		ufs_qcom_hosts[hba->dev->id] = host;
@@ -1520,56 +1555,6 @@ static int ufs_qcom_common_init(struct ufs_hba *hba)
 				__func__, err);
 
 	return 0;
-}
-
-/**
- * ufs_qcom_init - bind phy with controller
- * @hba: host controller instance
- *
- * Binds PHY with controller and powers up PHY enabling clocks
- * and regulators.
- *
- * Return: -EPROBE_DEFER if binding fails, returns negative error
- * on phy power up failure and returns zero on success.
- */
-static int ufs_qcom_init(struct ufs_hba *hba)
-{
-	struct device *dev = hba->dev;
-	struct ufs_qcom_host *host;
-	int err;
-
-	err = ufs_qcom_common_init(hba);
-	if (err)
-		goto out_variant_clear;
-
-	host = ufshcd_get_variant(hba);
-	/* Fire up the reset controller. Failure here is non-fatal. */
-	host->rcdev.of_node = dev->of_node;
-	host->rcdev.ops = &ufs_qcom_reset_ops;
-	host->rcdev.owner = dev->driver->owner;
-	host->rcdev.nr_resets = 1;
-	err = devm_reset_controller_register(dev, &host->rcdev);
-	if (err)
-		dev_warn(dev, "Failed to register reset controller\n");
-
-	if (!has_acpi_companion(dev)) {
-		host->generic_phy = devm_phy_get(dev, "ufsphy");
-		if (IS_ERR(host->generic_phy)) {
-			err = dev_err_probe(dev, PTR_ERR(host->generic_phy), "Failed to get PHY\n");
-			goto out_variant_clear;
-		}
-	}
-
-	err = ufs_qcom_init_lane_clks(host);
-	if (err)
-		goto out_variant_clear;
-
-	ufs_qcom_set_caps(hba);
-	ufs_qcom_advertise_quirks(hba);
-
-	ufs_qcom_setup_clocks(hba, true, POST_CHANGE);
-
-	return 0;
 
 out_variant_clear:
 	ufshcd_set_variant(hba, NULL);
@@ -1588,20 +1573,47 @@ static void ufs_qcom_exit(struct ufs_hba *hba)
 
 static int ufs_qcom_fmr_init(struct ufs_hba *hba)
 {
+	struct device *dev = hba->dev;
+	struct ufs_qcom_host *host;
 	int err;
 
-	err = ufs_qcom_common_init(hba);
+	host = devm_kzalloc(dev, sizeof(*host), GFP_KERNEL);
+	if (!host) {
+		dev_err(dev, "%s: no memory for qcom ufs host\n", __func__);
+		return -ENOMEM;
+	}
+
+	/* Make a two way bind between the qcom host and the hba */
+	host->hba = hba;
+	ufshcd_set_variant(hba, host);
+
+	ufs_qcom_get_controller_revision(hba, &host->hw_ver.major,
+		&host->hw_ver.minor, &host->hw_ver.step);
+
+	if (hba->dev->id < MAX_UFS_QCOM_HOSTS)
+		ufs_qcom_hosts[hba->dev->id] = host;
+
+	err = ufs_qcom_ice_init(host);
 	if (err)
 		goto out_variant_clear;
 
+	ufs_qcom_get_default_testbus_cfg(host);
+	err = ufs_qcom_testbus_config(host);
+	if (err)
+		/* Failure is non-fatal */
+		dev_warn(dev, "%s: failed to configure the testbus %d\n",
+				__func__, err);
+
+	ufs_qcom_advertise_quirks(hba);
+	hba->quirks &= ~UFSHCD_QUIRK_REINIT_AFTER_MAX_GEAR_SWITCH;
+
 	hba->caps |= UFSHCD_CAP_WB_EN;
-	hba->quirks |= UFSHCD_QUIRK_BROKEN_AUTO_HIBERN8;
+	ufs_qcom_set_host_params(hba);
 
 	return 0;
 
 out_variant_clear:
 	ufshcd_set_variant(hba, NULL);
-
 	return err;
 }
 
@@ -2031,6 +2043,21 @@ static int ufs_qcom_device_reset(struct ufs_hba *hba)
 	return 0;
 }
 
+static int ufs_qcom_fmr_device_reset(struct ufs_hba *hba)
+{
+        struct ufs_qcom_host *host = ufshcd_get_variant(hba);
+        static bool is_boot = true;
+
+        if (!is_boot) {
+                pm_runtime_put_sync(hba->dev);
+                pm_runtime_get_sync(hba->dev);
+        } else {
+                is_boot = false;
+        }
+
+        return 0;
+}
+
 #if IS_ENABLED(CONFIG_DEVFREQ_GOV_SIMPLE_ONDEMAND)
 static void ufs_qcom_config_scaling_param(struct ufs_hba *hba,
 					struct devfreq_dev_profile *p,
@@ -2314,6 +2341,7 @@ static const struct ufs_hba_variant_ops ufs_hba_qcom_fmr_vops = {
 	.suspend		= ufs_qcom_fmr_suspend,
 	.resume			= ufs_qcom_fmr_resume,
 	.dbg_register_dump	= ufs_qcom_dump_dbg_regs,
+	.device_reset		= ufs_qcom_fmr_device_reset,
 	.config_scaling_param   = ufs_qcom_config_scaling_param,
 	.program_key		= ufs_qcom_ice_program_key,
 	.derive_sw_secret	= ufs_qcom_ice_derive_sw_secret,
-- 
2.34.1

